test/zimfile.cpp

```diff
@@ -73,13 +65,15 @@ emptyZimFileContent()
   return content;
 }
 
-void
-generateEmptyZimFile(int fd)
+class TempZimFile : public TempFile
 {
-  const std::string c = emptyZimFileContent();
-  write(fd, &c[0], c.size());
-  close(fd);
-}
+public:
+  TempZimFile(const char* name, const std::string& content)
+    : TempFile(name)
+  {
+    write(fd(), &content[0], content.size());
+  }
+};
```

Comment:
I would really use a free function here as suggested by @MiguelRocha. The functionality of TempZimFile is just to create a temp file and write in it, no need to create another class.

Reply:
Done. As a result lines that used to look like

```
const TempZimFile tmpfile("invalid_zim_file", zimfileContent);
```

now read

```
const auto tmpfile = makeTempFile("invalid_zim_file", zimfileContent);
```

which IMHO is less readable.

Other than that, in this iteration I cleaned up the PR a little by moving the `makeTempFile()` function to the earliest commit where it could appear and removing the intermediate commit introducing the former `TempZimFile` class.

Reply:
I'm not sure we need the `const` here. But if you really want it, you can change the signature to
`std::unique_ptr<const TempFile> makeTempFile(const char* name, const std::string& content)`
and use it this way :
`auto tmpfile = makeTempFile("invalid_zim_file", zimfileContent);`

And IMHO, is more readable because you keep the variable name on the left, as most initialization occur in cpp.

I also wonder about the need to use a `unique_ptr`. A plain return by value is enough here.
But anyway, I will not block for that.

Reply:
@mgautierfr 

> I'm not sure we need the `const` here. But if you really want it, you can change the signature to
`std::unique_ptr<const TempFile> makeTempFile(const char* name, const std::string& content)`

`const std::unique_ptr<T> ptr1` is not the same as `std::unique_ptr<const T> ptr2` - the former **cannot** be `reset()`, move-assigned to another variable or passed/returned by move-copy to another function. Therefore the lifetime of the object referred to by `ptr1` can be guaranteed to exactly extend to the lifetime of `ptr1`. Overall I am a big fan of defensive programming and use `const` wherever I can (I would even prefer if variables/types in C++ were immutable/`const` by default and one had to explicitly mark them as mutable/`nonconst`). Regarding using `std::unique_ptr<const TempFile>` as the return type of `makeTempFile()`: I had considered that but decided not to abuse `const` in this case, since `TempFile` is an immutable type anyway (its data members are encapsulated and all of its member functions are `const`).

> I also wonder about the need to use a `unique_ptr`. A plain return by value is enough here.

`TempFile` is a non-copyable type.